### ISSUE TITLE
shared_ptr idle deletion

### ISSUE THREAD

**Issue by ohhmm** [microsoft/STL#1516]
std::shared_ptr  use is scope-agnostic which allow us to dispatch its object deletion on an idle priority thread.

**Comment by StephanTLavavej** [comment]
Is there a question here?

**Comment by ohhmm** [comment]
Wouldn't it be great to have it in a way in a standard?

**Comment by AlexGuteniev** [comment]
I don't think so, such a facility is not general purpose enough.

**Comment by ohhmm** [comment]
Please, @AlexGuteniev,  

Could you describe a case where it may be not an option?

Thanks,
Serg

**Comment by miscco** [comment]
I think "idle priority thread" is already something that cannot be guaranteed.

What the standard needs are hard guarantees that are always fulfilled on essentially all platforms. Just because there might be *some* platform that could optimize behavior under *some* conditions, that is nothing the standard could endorse.

It is the same story like "unordered_map" is so terrible. It is not, it is just that the guarantees it provides are mostly unneeded. However, there are still valid use cases so just dropping them is a no go.

If you really have sufficient knowledge about *your* specific system and you *really* need to squeeze out the last cycles, you must write your own specialized implementation.

**Comment by miscco** [comment]
Also you could easily imagine some terrible code being written like:

```cpp
std::weak_ptr<Meow> weakPtr;
{
    std::shared_ptr<Meow> sharedPtr{};
    weakPtr = sharedPtr;
}

if (auto locked = weakPtr.lock) {
    
}
``` 

I am not sure how you would guarantee that `sharedPtr` is destroyed in time

**Comment by ohhmm** [comment]
> I think "idle priority thread" is already something that cannot be guaranteed.
> 
> What the standard needs are hard guarantees that are always fulfilled on essentially all platforms. Just because there might be _some_ platform that could optimize behavior under _some_ conditions, that is nothing the standard could endorse.
> 
> It is the same story like "unordered_map" is so terrible. It is not, it is just that the guarantees it provides are mostly unneeded. However, there are still valid use cases so just dropping them is a no go.
> 
> If you really have sufficient knowledge about _your_ specific system and you _really_ need to squeeze out the last cycles, you must write your own specialized implementation.

Just like SMP. But we have std::thread and std::async in standard. I think that if OS has multi-thread then it has idle priority. Please, correct me if you know cases of multi-threaded OS which has no idle priority scheduling.

**Comment by ohhmm** [comment]
> Also you could easily imagine some terrible code being written like:
> 
> ```c++
> std::weak_ptr<Meow> weakPtr;
> {
>     std::shared_ptr<Meow> sharedPtr{};
>     weakPtr = sharedPtr;
> }
> 
> if (auto locked = weakPtr.lock()) {
>     // you do not want to end up here
> }
> ```
> 
> I am not sure how you would guarantee that `sharedPtr` is destroyed in time

Once object dispatched for deletion, it considered as deleted in current semantics. This code would work correctly.

**Comment by MikeGitb** [comment]
@ohhmm: 

The way `shared_ptr` is currently specified, it gives me hard guarantees, when an object is destroyed, which can be important, because the destructor may have side-effects (like unlockign a mutex or releasing other resources) and at the very least it has an effect on the available memory on my system. If you shedule deletion on an idle thread you a) have no control over when the destructor is run and b) you might actually have races if the destructor modifes some state that is also accessed from another thread (if nothing else, it is quite likely that the destructor accesses a global allocator).

That isn't to say that there aren't many valid usecases for this style of garbage collection. It just doesn't fit with the specification of `std::shared_ptr`, so this is nothing the STL maintainers can implement, even if they thought it was a good idea.  

If you think the standard library should provide a type with those semantics (it would have to be a separate type, because it would be a breaking change for shared_ptr), you need to write a paper and convince the committee of it (for the record I'd say the chances are rather slim, but I've never been to a meeting, so what do I know).

**Comment by ohhmm** [comment]
Thanks @MikeGitb,

It makes sense. 
I wouldn't tell about hard guaranties though because shared_ptr is multithreaded and another thread may be suspended by OS.
But overall your opinion seem inline with potential committee opinion.
I was asking  it here to see if it gets any support by Microsoft.
Usually distinct person would not write a proposal to committee. 
It guided by big companies.
I am trying to seed this idea.

Thanks,
Serg

**Comment by KubaO** [comment]
> dispatch its object deletion on an idle priority thread

This requires that *all* destructors for classes used with `std::unique_ptr` are thread-safe. Generally speaking, this is unlikely to be the case. I think the only way to make it work would be to add a special trait that enables a given type to be processed this way. But this is all very much outside of the scope of standard C++, I think, and wouldn't belong in STL according to goals/non-goals...?

**Comment by ohhmm** [comment]
I seen profiling results where shared_ptr destruction takes about ~30% 
Rust has asynchronous destructors feature already BTW.
It is totally safe and it totally makes sense in all cases.

**Comment by KubaO** [comment]
> It is totally safe and it totally makes sense in all cases.

It absolutely isn’t totally safe if you didn’t design the destructors in the entire inheritance hierarchy for a given class to be thread-safe. It’s trivial to break it so the assertion that it’s “totally safe” holds no water. Yes, shared ptr is thread safe by itself. That’s a requirement but not sufficient. The objects being destroyed must support being destroyed in another thread, and that is absolutely not generally the case nor guaranteed nor can be depended upon. At **best** it can be an opt-in for some class hierarchies.

Even if you run the destructor immediately but defer the memory release till later, you may be killing performance of memory allocators that were designed to leverage thread locality.

So, generally it’s a no-go, but if you wanted to implement it for some classes in your project, it’s like a page of code and ten pages of tests - no big deal.

**Comment by ohhmm** [comment]
I am still 100% sure it is totally safe.

**Comment by AdamBucior** [comment]
> I am still 100% sure it is totally safe.

No it's not. There are objects that need to be created and destroyed on the same thread, for example a window in Win32 API).

**Comment by sylveon** [comment]
Another example is anything XAML-related (in UWP or the upcoming WinUI 3) must only be accessed from the GUI thread. Destructors performing cleanup from a background thread would irrecoverably crash the app.

Unlocking a SRWLOCK must be done on the same thread that locked it - so if the destructor unlocks one, you're done too (binary semaphores tend to be the only thread synchronization primitive which support locking on a certain thread and then unlocking on another).

If the destructor accesses global data, this change can introduce new data races silently.

While your idea is a pretty good idea in theory, a generic-purpose library like the STL can't support it because a lot of code it supports relies on destructors being ran on the same thread. IIRC Rust's async destructors don't necessarily specify destruction on a background thread - it is very much possible to be asynchronous on a single thread.

**Comment by KubaO** [comment]
Another case in point: most UI-related bits in Cocoa can’t be touched from any thread but the first one in the process, `QWidget` hierarchy in Qt - same, also any `QObject` that runs timers can’t be destroyed in other threads, and so on. And that stuff won’t just complain for no reason, it crashes very readily to make that point, and there definitely are cases when you hold a `QWidget` or `QObject` in a `shared_ptr`.

But some way of this being opt-in would be cool - also for unique_ptr. But I imagine that there would need then to be a fallback in all memory allocations to wait on failure till all deferred deletion queues get emptied and retry. Gets rather complicated…

**Comment by ohhmm** [comment]
These examples are are using shared_ptr where reference applicable. I think its not a real world example. 

Can anyone provide real working software which use shared_ptr which checked to become broken because of shared_ptr objects deletion dispatched to idle priority?

I don't think you can find such an example.

**Comment by ohhmm** [comment]
If you use shared_ptr then you don't know which thread deletes the object by standard. Which makes it reliable to dispatch.

**Comment by miscco** [comment]
I believe I gavew you an example that clearly shows that this *not* correct. 

Also in most cases this would also be a huge performance grave. Imagine a shared_ptr that holds a trivially destructible struct. You just created a huge mess without any benefit. 

Again, if you want a certain specific use case to behave adesired you need to write your own tool. The standard cannot help you there.

**Comment by AlexGuteniev** [comment]
> If you use shared_ptr then you don't know which thread deletes the object by standard. Which makes it reliable to dispatch.

Not necessarily. If I don't share instances of `shared_ptr` between threads, I know.
Most likely example is scope guard.

Another thing is I rely on deletion of last `shared_ptr` reference to do something, for example joining threads.

A lot of my code will break.
Some more will have degraded performance.

**Comment by ohhmm** [comment]
> > If you use shared_ptr then you don't know which thread deletes the object by standard. Which makes it reliable to dispatch.
> 
> Not necessarily. If I don't share instances of `shared_ptr` between threads, I know.
> Most likely example is scope guard.
> 
> Another thing is I rely on deletion of last `shared_ptr` reference to do something, for example joining threads.
> 
> A lot of my code will break.
> Some more will have degraded performance.

Thanks @AlexGuteniev,

Did you ever see 100% load during real use? 
There is always a way for idle processing in modern PC. 
Process exiting delay just moved from software runtime to the end of the runtime.
Still I'm not sure it be slower.
Currently it just your words versus my.
And I am 100% sure.
I would even bet it to see a real software which would degrade performance or show new introduced bug because of this feature.

Thanks,
Serg

**Comment by AlexGuteniev** [comment]
First, mostly important is still correctness issue. As long as it will become incorrect for many cases, the change cannot be made. For example, I join threads or disconnect sockets or do anything like this in destructors. (Have you tried to use `asio` without `shared_ptr`? Or can you imaging working with `asio` when per-connection entities deletions are arbitrarily delayed?)

Or, haven't ever used "lazy scope guard" based on `shared_ptr` deleter?

Maybe you are right in that the perfromance problem does not apply to `shared_ptr` though. My SW is able to utilize 100% CPU, but such software doesn't use `shared_ptr` for anything performance critical. Performance-critical code minimizes allocations at all. So it is primarily correctness concern.

**Comment by AlexGuteniev** [comment]
One more point against such a facility. Passing a deleter to the current `shared_pt` that would delete `shared_ptr` in own thread is literally one-liner (a lambda with `std:;async`). But if it is implemented by default, there's no way to opt out.

**Comment by KubaO** [comment]
All it takes to break things is if the object being destructed modifying a shared data structure in the destructor - say removing itself from an observer list, or from an index. Instant undefined behavior. I mean, this is obvious stuff, right?

I think that there's some merit to the general idea of running memory deallocation (after the destructor has finished) asynchronously, to save time, but this either 1. infects every allocation with an out-of-memory retry and additional synchronization work, in case deallocations were queued but not finished yet, or 2. requires participation from the allocator which, in case of out-of-memory condition, would have to take up deallocation work.

> If you use `shared_ptr` then you don't know which thread deletes the object by standard. Which makes it reliable to dispatch.

Can you cite anything in the standard that implies that? Because if so, I have large swaths of code where I need to stop using `shared_ptr` right away... The way I see it, you only don't know which thread deletes the object if you share the object between threads, and in a way where the "known" thread loses all references while some other threads still hold references. `shared_ptr` is designed to *cope* with multi-threaded use, but it doesn't in any way mandate it. I'd say that a much more worthwhile C++ library improvement (which has been shot down, IIRC), would have been to have a single-threaded `shared_ptr` without atomic reference counting - that can improve the very common use cases. But a decision was made that this was too specific of an optimization to make it worth the trouble. `shared_ptr` is supposed to be as widely applicable as possible, and I think it does that job just fine!

**Comment by CaseyCarter** [comment]
We're not interested in implementing or evangelizing this feature. As others have mentioned, this would create differences in observable behavior, making it non-conforming.

**Comment by ohhmm** [comment]
> First, mostly important is still correctness issue. As long as it will become incorrect for many cases, the change cannot be made. For example, I join threads or disconnect sockets or do anything like this in destructors. (Have you tried to use `asio` without `shared_ptr`? Or can you imaging working with `asio` when per-connection entities deletions are arbitrarily delayed?)
> 
> Or, haven't ever used "lazy scope guard" based on `shared_ptr` deleter?
> 
> Maybe you are right in that the perfromance problem does not apply to `shared_ptr` though. My SW is able to utilize 100% CPU, but such software doesn't use `shared_ptr` for anything performance critical. Performance-critical code minimizes allocations at all. So it is primarily correctness concern.

Thanks @AlexGuteniev,

For being professionally polite and being driven by reason.
Yes, I did use boost::asio. 
It wouldn't be broken with this. You can try it.

In the question I typed "in a way" which implies possibility to implement it as new smart pointer type. 

I understand that sometimes developer may store a lock inside a shared_ptr.
This case wouldn't be neither broken nor degrade in performance unless we had 100% load for all hw cores blocking processing of idle priority tasks.
There is no real world broken case.
shared_ptr made not for scoping by design (where the next operator waits for deletion of an object that is used for the scoping). 
scoped_ptr made for scoping.

No good reason to create std::thread to instantly rely on artifact of its deletion.
If you do std::async oneliner then you rely on artifact of its execution not an artifact of its deletion.
In the question I typed "in a way" which implies possibility to implement it as new smart pointer type anyway. 

I still think that shared_ptr modification has more reason then doing new ptr wrapper.
Because fixing esoterically couple cases of already buggy code is better then substituting millions of current safe usages and people are frightened to do that.
So in terms of saving world's time I would stick with idling shared_ptr wrapped object deletion.

Thanks,
Serg

**Comment by ohhmm** [comment]
Hi @KubaO,

> All it takes to break things is if the object being destructed modifying a shared data structure in the destructor - say removing itself from an observer list, or from an index. Instant undefined behavior. I mean, this is obvious stuff, right?
> 
> I think that there's some merit to the general idea of running memory deallocation (after the destructor has finished) asynchronously, to save time, but this either 1. infects every allocation with an out-of-memory retry and additional synchronization work, in case deallocations were queued but not finished yet, or 2. requires participation from the allocator which, in case of out-of-memory condition, would have to take up deallocation work.
> 
> > If you use `shared_ptr` then you don't know which thread deletes the object by standard. Which makes it reliable to dispatch.
> 
> Can you cite anything in the standard that implies that? Because if so, I have large swaths of code where I need to stop using `shared_ptr` right away... The way I see it, you only don't know which thread deletes the object if you share the object between threads, and in a way where the "known" thread loses all references while some other threads still hold references. `shared_ptr` is designed to _cope_ with multi-threaded use, but it doesn't in any way mandate it. I haven't dug deep into the standard to confirm that, but every implementation I'm aware of behaves in the expected way and adds no thread safety requirements where there would have been none were `shared_ptr` not in use.
> 
> A much more worthwhile C++ library improvement would have been to have a single-threaded `shared_ptr` without atomic reference counting - that can improve the very common use cases. But a decision was made that this was too specific of an optimization to make it worth the trouble. `shared_ptr` is supposed to be as widely applicable as possible, and I think it does that job just fine!
> 
> The way I read it, by the time `~shared_ptr` returns, the managed object must have been deleted. So this "optimization" can't happen without a fundamental and breaking change to the C++ library, so it's a no go, and this issue should be closed. It's not meritless or foolish, just not something that can be done while staying within confines of the standard.


Please, let me understand your statement clearly.
We're deleting objects that wrapped inside shared_ptr.
All its behavior remains except one single point: closing of the shared_ptr group object.
There is literally no difference in behavior from the standard point of view but priority of thread which calls wrapped object destructor.
So the only thing that changed from the standard point of view is the single thing:
    Priority of a thread which does deletion is down to idle.
There is literally no any other difference from the Committee C++ standard point of view.

The best people arguments here is about potential performance degradation due to this.
My argument for these opinions is that there is impossible to find real product or library software which would degrade.
It is clear to me because performance gain is 100% sure much bigger.

If its still unclear why the priority is the only thing changed then I may rephrase one of my previous statements:
shared_ptr is for objects when it is unknown which thread deletes it first. Which makes unreliable to rely on any kinds of scoping on the shared_ptr scoping basis.
shared_ptr scoping for every particular thread is irrelevant to the object life time => non thread rely on the shared_ptr life time.

Thanks,
Serg

**Comment by AlexGuteniev** [comment]
Releasing mutex in a different thread is undefined behavior.
Some mutex implementations track owning thread and using per-thread structures.
`CoUninitialize` in different thread spreads havoc.

What you actually want I guess is not spending time on deallocating memory and not caring about object lifetime. Garbage collection is more efficient solution for this than reference-counted smart pointer, since it does not involve atomic reference counting. Garbage collection also means not using RAII, that is no business logic in destructor, and no `CoInitialize` or mutex lock RAII wrappers. This way it works.

Unfortunately, it probably means you'll have to switch to a language with garbage collection, since C++ mentions "minimal support for garbage collection", but does not actually support it.

----

Note also that if you are still convinced that async freeing of `shared_ptr` is valuable, and think you can convince others, you'll have to implement it on your own. You can fork this STL implementation, but pull requests will not be accepted back. This STL implementation has set non-Standard extensions as non-goal. And to make your proposal Standard, you'll need to propose it for Standard, not here:

> If you're proposing a feature to WG21 (the C++ Standardization Committee), you're welcome (and encouraged!) to use our code as a base for a proof-of-concept implementation. These non-goals simply mean that we're unable to consider pull requests for a proposed feature until it has been voted into a Working Paper. After that happens, we'll be delighted to review a production-ready pull request.

**Comment by ohhmm** [comment]
Thanks @AlexGuteniev,

Lets also consider ifdef-conditional feature for this.
If we had this conditional feature then we could use it in some cases and check it for other cases.
Would that `_SHARED_PTR_IDLE_PRIORITY_DELETION` be good?

Thanks,
Serg
